src/core/src/op/loop.cpp

```diff
@@ -191,47 +191,100 @@ void op::v5::Loop::validate_and_infer_types() {
     // Body
     m_bodies[0]->validate_nodes_and_infer_types();
 
+    // Port map processing: output -> input
+    std::map<uint64_t, uint64_t> back_edges;
+    for (const auto& desc : get_input_descriptions()) {
+        if (const auto& merged_desc = std::dynamic_pointer_cast<opset5::Loop::MergedInputDescription>(desc)) {
+            back_edges[merged_desc->m_body_value_index] = merged_desc->m_body_parameter_index;
+        }
+    }
     // Output
-    for (const auto& output_description : m_output_descriptions[0]) {
-        auto index = output_description->m_output_index;
-
-        auto body_value = m_bodies[0]->get_results().at(output_description->m_body_value_index)->input_value(0);
-
-        if (auto concat_output_description =
-                ov::as_type_ptr<v0::TensorIterator::ConcatOutputDescription>(output_description)) {
-            const auto& body_value_partial_shape = body_value.get_partial_shape();
-            auto out_shape = body_value_partial_shape;
-            if (zero_number_of_iter) {
-                out_shape = ov::PartialShape{0};
-            } else if (out_shape.rank().is_static()) {
-                const auto axis = ngraph::normalize_axis(this, concat_output_description->m_axis, out_shape.rank());
-                const auto rank = out_shape.rank().get_length();
-                if (rank == 0) {
-                    out_shape = ov::PartialShape{1};
+    bool need_reinvalidate = false;
+    // 2 pass: if output shape is not same with input in a back-edge, first pass will update the
+    //   input shape, second pass will re-compute the output shape using the changed input shape
+    // 1 pass: otherwise
+    for (int i = 0; i < 2; i++) {
```

Comment:
Could you elaborate why you feel like two times is enough? Do you attribute it to single change of input shape due to back edge situation? Could we have multiple shape changes? @evolosen could you follow up on this?

Reply:
Parameter shape may be changed because the output shape is not the same with parameter shape in the back edge, the first pass will change all parameter shape if needed, the second pass will use the fixed parameters' shape to compute the output shape.
If we have multiple back edges caused the corresponding input parameters shape changed, we could have multiple shape changes.
I made [a model](https://github.com/luo-cheng2021/openvino/blob/e2cb6ab795a3d9a8c8de3b77a37c81807c99b2ec/src/core/tests/type_prop/loop.cpp#L1194) which has multiple shape changes.